### PR TITLE
Avoiding TOCTOU vulnerability in ProtobufReaderWriter class

### PR DESCRIPTION
**Pull request by tomaszrybicki**
Signed-off-by: Tomasz Rybicki <tomasz.rybicki@intel.com>

### CODE REVIEW

File: source/octf/node/NodeBase.cpp [hunk 1]
```diff
@@ -60,16 +66,21 @@ bool NodeBase::readSettings() {
...
+            ProtobufReaderWriter rw(
+                    getFrameworkConfiguration().getNodeSettingsFilePath(
+                            getNodePath()));
 
-        if (rw.write(*m_config)) {
-            return true;
-        } else {
-            log::cerr << "Cannot write configuration for file "
-                      << rw.getFilePath() << std::endl;
+            if (rw.write(*m_config)) {
```
Comment (mariuszbarczak) [inline source/octf/node/NodeBase.cpp]: 👍 It's good idea to try catch in write path which is involved during de-initialization. Thus allows to perform another de-initialization activities.

File: source/octf/node/NodeBase.cpp [hunk 2]
```diff
@@ -41,16 +41,22 @@ void NodeBase::deinitCommon() {
 
 bool NodeBase::readSettings() {
     if (m_config) {
-        ProtobufReaderWriter rw(
-                getFrameworkConfiguration().getNodeSettingsFilePath(
-                        getNodePath()));
+        try {
```
Comment (mariuszbarczak) [inline source/octf/node/NodeBase.cpp]: Impossibility of reading previous settings it's critical, so I'd throw exception to stop initialization immediately.

File: source/octf/utils/ProtobufReaderWriter.cpp [hunk 3]
```diff
@@ -31,47 +35,57 @@ ProtobufReaderWriter::ProtobufReaderWriter(const std::string &filePath)
...
+    status = google::protobuf::util::JsonStringToMessage(fileContent, &message,
                                                          opts);
 
     return status.ok();
 }
 
 bool ProtobufReaderWriter::write(const google::protobuf::Message &message) {
-    // 1. First generate string with configuration
+    if (m_writeFd == -1) {
+        return false;
```
Comment (mariuszbarczak) [inline source/octf/utils/ProtobufReaderWriter.cpp]: Please make consistent behavior for read and write paths when file descriptor is not available.

File: source/octf/utils/ProtobufReaderWriter.cpp [hunk 4]
```diff
@@ -124,62 +136,82 @@ bool ProtobufReaderWriter::isFileAvailable(std::fstream::openmode access) {
...
 
-bool ProtobufReaderWriter::remove() {
-    bool result;
-    try {
-        if (isFileAvailable()) {
-            result = ::remove(m_filePath.c_str());
-            return (result == 0);
+void ProtobufReaderWriter::openFile() {
+    // File descriptor for writing
+    int writeFd = ::open(m_filePath.c_str(),
```
Comment (mariuszbarczak) [inline source/octf/utils/ProtobufReaderWriter.cpp]: Does it truncate file upon opening?

Comment (tomaszrybicki) [reply source/octf/utils/ProtobufReaderWriter.cpp]: File should not be truncated here.

**Pull request merged**